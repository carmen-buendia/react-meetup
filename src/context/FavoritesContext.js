import React, { createContext, useContext, useState, useEffect } from "react";

const FavoritesContext = createContext();

export function FavoritesProvider({ children }) {
  const [favorites, setFavorites] = useState([]);
  const [meetups, setMeetups] = useState([]);

  useEffect(() => {
    const fetchMeetups = async () => {
      try {
        const response = await fetch("/data.json");
        const data = await response.json();
        setMeetups(data);
      } catch (error) {
        console.error("Error loading the meetups", error);
      }
    };

    fetchMeetups();
  }, []);

  const addFavorite = (item) => {
    setFavorites((prevFavorites) => [...prevFavorites, item]);
  };

  const removeFavorite = (itemId) => {
    setFavorites((prevFavorites) =>
      prevFavorites.filter((item) => item.id !== itemId)
    );
  };

  return (
    <FavoritesContext.Provider
      value={{ favorites, addFavorite, removeFavorite, meetups }}
    >
      {children}
    </FavoritesContext.Provider>
  );
}

// Hook para usar el contexto
export const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error("useFavorites must be used within a FavoritesProvider");
  }
  return context;
};
