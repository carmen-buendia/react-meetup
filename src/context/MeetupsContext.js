import React, { createContext, useContext, useState } from "react";

const MeetupsContext = createContext();

export const MeetupsProvider = ({ children }) => {
  const [meetup, setMeetup] = useState([]);

  const addMeetup = (newMeetup) => {
    setMeetup((meetup) => [...meetup, newMeetup]);
  };

  return (
    <MeetupsContext.Provider value={{ meetup, addMeetup }}>
      {children}
    </MeetupsContext.Provider>
  );
};

export const useMeetup = () => {
  return useContext(MeetupsContext);
};
