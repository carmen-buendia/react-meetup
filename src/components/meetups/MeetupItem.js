import React, { useState, useEffect, useContext } from "react";
import { FavoritesContext } from "../../context/FavoritesContext";
import classes from "./MeetupItem.module.css";
import Card from "../ui/Card";

export default function MeetupItem({ meetup }) {
  const { favorites, addFavorite, removeFavorite } =
    useContext(FavoritesContext);
  const [isFavorite, setIsFavorite] = useState(false);

  useEffect(() => {
    setIsFavorite(favorites.some((fav) => fav.id === meetup.id));
  }, [favorites, meetup.id]);

  const toggleFavoriteHandler = () => {
    if (isFavorite) {
      removeFavorite(meetup.id);
    } else {
      addFavorite(meetup);
    }
    setIsFavorite(!isFavorite);
  };

  return (
    <li className={classes.item} data-test="meet-up-item">
      <Card>
        <div className={classes.image}>
          <img src={meetup.image} alt={meetup.title} />
        </div>
        <div className={classes.content}>
          <h3>{meetup.title}</h3>
          <address>{meetup.address}</address>
          <p>{meetup.description}</p>
        </div>
        <div className={classes.actions}>
          <button onClick={toggleFavoriteHandler}>
            {isFavorite ? "Remove from Favorites" : "Add to Favorites"}
          </button>
        </div>
      </Card>
    </li>
  );
}
