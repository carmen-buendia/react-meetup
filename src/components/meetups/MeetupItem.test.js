/* eslint-disable testing-library/no-debugging-utils */
import React from "react";
import { mount } from "enzyme";
import MeetupItem from "./MeetupItem";
import { FavoritesContext } from "../../context/FavoritesContext";

const mockMeetup = {
  id: "m1",
  title: "Test Meetup",
  image: "https://example.com/image.jpg",
  address: "123 Test Street",
  description: "This is a test description.",
};

const mockAddFavorite = jest.fn();
const mockRemoveFavorite = jest.fn();

describe("<MeetupItem />", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <FavoritesContext.Provider
        value={{
          favorites: [],
          addFavorite: mockAddFavorite,
          removeFavorite: mockRemoveFavorite,
        }}
      >
        <MeetupItem meetup={mockMeetup} />
      </FavoritesContext.Provider>
    );
  });

  test("renders without crashing", () => {
    expect(wrapper.exists()).toBe(true);
  });

  test("renders the correct title", () => {
    expect(wrapper.find("h3").text()).toBe(mockMeetup.title);
  });

  test("renders the correct image", () => {
    const img = wrapper.find("img");
    expect(img.prop("src")).toBe(mockMeetup.image);
    expect(img.prop("alt")).toBe(mockMeetup.title);
  });

  test("renders the correct address", () => {
    expect(wrapper.find("address").text()).toBe(mockMeetup.address);
  });

  test("renders the correct description", () => {
    expect(wrapper.find("p").text()).toBe(mockMeetup.description);
  });

  test("button toggles correctly for favorites", () => {
    const button = wrapper.find("button");

    expect(button.text()).toBe("Add to Favorites");
    button.simulate("click");

    expect(mockAddFavorite).toHaveBeenCalledWith(mockMeetup);
    expect(button.text()).toBe("Remove from Favorites");

    button.simulate("click");
    expect(mockRemoveFavorite).toHaveBeenCalledWith(mockMeetup.id);
    expect(button.text()).toBe("Add to Favorites");
  });

  test("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
