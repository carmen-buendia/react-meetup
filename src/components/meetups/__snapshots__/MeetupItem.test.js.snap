// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MeetupItem /> matches snapshot 1`] = `ReactWrapper {}`;
